### ISSUE TITLE
[MBUILDCACHE-33] Support remote cache credentials from environment variables 

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-build-cache-extension#292]
**[Alexander Ashitkin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alex_ashitkin)** opened **[MBUILDCACHE-33](https://issues.apache.org/jira/browse/MBUILDCACHE-33?redirect=false)** and commented

In my current environment settings.xml are managed by a build team which is not allowing any modification because the same build service is used by all teams. Atop of that, maven build runs in a fresh container which doesn't have any credentials injected for security reasons. Because of that cache cannot read/deploy build artifacts to an authenticated http server. Still, our build service allows to inject credentials from environment variables into build container. Need to support cache setup without settings.xml by injecting environment variables:
* MAVEN_BUILD_CACHE_DIRECT_CONNECT
* MAVEN_BUILD_CACHE_USER
* MAVEN_BUILD_CACHE_PASSWORD
* MAVEN_BUILD_CACHE_PROXY_USER
* MAVEN_BUILD_CACHE_PROXY_PASSWORD

---

**Remote Links:**
- [GitHub Pull Request #35
](https://github.com/apache/maven-build-cache-extension/pull/35)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Here is a similar issue: https://issues.apache.org/jira/browse/MNG-7050

**Comment by jira-importer** [comment]
**[Alexander Ashitkin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alex_ashitkin)** commented

It is not similar. Let's narrow down case to a build which needs secrets in a container on Kubernetes.
* Option1: image with embedded settings.xml - no go, insecure to publish credentials into binary repo and complicates maintenance in case of credentials expiry
* Option2: Mount secrets.xml from a shared volume - again no go, as it requires the same - store secrets in some 3rd storage where they could leak
* Option 3: Mount settings xml from secrets as a volume - again inconvenient as it mixes 2 orthogonal problems - maven setup management and  credentials management. Imagine need to change mirror or add proxy. What to do? Edit secrets?

Usage of settings xml in build environment is either insecure, either add complexity. Moreover, any access to credentials in settings.xml at build  time is a potential security breach, as they could be easily leaked by redirecting repository call to a sniffer. So realistically, there is no good case to keep anything but personal credentials in settings.xml on own desktop. 

Now, speaking of secure setup - the only way in managed environment to control credentials is to use secrets service and each major cloud platform provides some form of such a service. Usually, credentials are mounted on container either like a volume (mainly for keys) or as env var - for individual secrets, like password.
Now, having typical case of container securely bootstrapped with injected secrets, there are just 2 options to use them in build:
1) Directly, as in this pr.
2) Indirectly, by resolving placeholders in config files from ${env.MY_VAR} 
3) Indirectly, by setting vm flags from env vars: -Dcache.password=${CACHE_PASSWORD}

Option 2 doesn't work in settings.xm. Even if it works it will require to store settings.xml if build environment is shared. Option 3 increases chances to leak credentials through logs if masked improperly or if command line is captured in tooling.
 
So basically, in container environment, settings.xml is a big security leak. In terms of convenience only env vars provide convenient and secure way to inject credentials in build container. Though it's possible to find working solution using one of the ways above, but again that's likely insecure or inconvenient.
Environment variable is the optimal way.

I hope it explains!

**Comment by jira-importer** [comment]
**[Guillaume Nodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnt)** commented

[Alexander Ashitkin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alex_ashitkin) I fail to see how this issue is specific to maven-build-cache-extension.  If the credentials mechanism in maven is somehow flawed, it should be fixed in maven, rather than for each specific plugin/extension I'd think.

I think a per-project `settings.xml` has been proposed.  Wouldn't that solve the problem ?

See https://issues.apache.org/jira/browse/MNG-5659

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

A per project settings does not make sense since you would expose credentials via SCM...

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

A couple of years ago I had the idea to completely abstract the credential provider similar to what Git or Subversion to. We provide the resource and/or id and the provider returns creds.

**Comment by jira-importer** [comment]
**[Guillaume Nodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnt)** commented

The settings do not have to be put in scm.  Of course that would be a bad idea if they contain credentials.

**Comment by jira-importer** [comment]
**[Alexander Ashitkin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=alex_ashitkin)** commented

[gnodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnodet) putting credentials in settings.xml is either insecure or inconvenient. Imagine large enterprise team with hundreds of projects. Imagine level of inconvenience to create hundreds of identical settings.xml just to put the same credentials hundreds of time there? Imagine how much pain is to change such credentials. Even not speaking of build containers in the old good Jenkins build secrets are normally injected to build with environment variables ([https://www.jenkins.io/doc/pipeline/steps/credentials-binding)](https://www.jenkins.io/doc/pipeline/steps/credentials-binding/). Convenient setup nowadays is to manage common maven configs in a shared settings.xml and manage repo credentials in a centralized secret manager. You're right the problem is not limited to the cache. Ideally environment variables should be supported in Maven core because. But at this point it is possible to solve it in cache at least which will allow to use cache in compliant manner in a properly secured build environments

**Comment by jira-importer** [comment]
**[Alexander Schwartz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ahus1)** commented

In a previous comment, it is stated that `${env.MY_VAR`} is not working for a settings.xml.

I've used it myself in this example https://github.com/ahus1/keycloak-dropwizard-integration/blob/master/cisettings.xml so I suppose this can be used in other setups as well.

I've been using this to have a per-project `settings.xml` checked into the repo next to the pom.xml, and keeping the secrets as environment variables. The project-specific settings can then be passed to maven on the CLI as `mvn --settings cisettings.xml ...`
